src/wasm/nxt_wasm.c

```diff
@@ -0,0 +1,262 @@
+        const char *name;
+        const char *val;
+
+        name = (const char *)(uint8_t *)rh + rh->fields[i].name_offs;
+        val = (const char *)(uint8_t *)rh + rh->fields[i].value_offs;
+
+        nxt_unit_response_add_field(ctx->req, name, rh->fields[i].name_len,
+                                    val, rh->fields[i].value_len);
+    }
+
+    nxt_unit_response_send(ctx->req);
+}
+
+
+void
+nxt_wasm_do_send_response(uint32_t offs, nxt_wasm_ctx_t *ctx)
+{
+    nxt_wasm_response_t      *resp;
+    nxt_unit_request_info_t  *req = ctx->req;
+
+    if (!nxt_unit_response_is_init(req)) {
+        nxt_unit_response_init(req, 200, 0, 0);
+    }
+
+    resp = (nxt_wasm_response_t *)(nxt_wasm_ctx.baddr + offs);
+
+    nxt_unit_response_write(req, (const char *)resp->data, resp->size);
+}
+
+
+static void
+nxt_wasm_request_handler(nxt_unit_request_info_t *req)
+{
+    size_t                     offset, read_bytes, content_sent, content_len;
+    ssize_t                    bytes_read;
+    nxt_unit_field_t           *sf, *sf_end;
+    nxt_unit_request_t         *r;
+    nxt_wasm_request_t         *wr;
+    nxt_wasm_http_hdr_field_t  *df;
+
+    NXT_WASM_DO_HOOK(NXT_WASM_FH_REQUEST_INIT);
+
+    wr = (nxt_wasm_request_t *)nxt_wasm_ctx.baddr;
+
+#define SET_REQ_MEMBER(dmember, smember) \
+    do { \
+        const char *str = nxt_unit_sptr_get(&r->smember); \
+        wr->dmember##_offs = offset; \
+        wr->dmember##_len = strlen(str); \
+        memcpy((uint8_t *)wr + offset, str, wr->dmember##_len + 1); \
+        offset += wr->dmember##_len + 1; \
+    } while (0)
+
+    r = req->request;
+    offset = sizeof(nxt_wasm_request_t)
+             + (r->fields_count * sizeof(nxt_wasm_http_hdr_field_t));
+
+    SET_REQ_MEMBER(path, path);
+    SET_REQ_MEMBER(method, method);
+    SET_REQ_MEMBER(version, version);
+    SET_REQ_MEMBER(query, query);
+    SET_REQ_MEMBER(remote, remote);
+    SET_REQ_MEMBER(local_addr, local_addr);
+    SET_REQ_MEMBER(local_port, local_port);
+    SET_REQ_MEMBER(server_name, server_name);
+#undef SET_REQ_MEMBER
+
+    df = wr->fields;
+    sf_end = r->fields + r->fields_count;
+    for (sf = r->fields; sf < sf_end; sf++) {
+        const char *name = nxt_unit_sptr_get(&sf->name);
+        const char *value = nxt_unit_sptr_get(&sf->value);
+
+        df->name_offs = offset;
+        df->name_len = strlen(name);
+        memcpy((uint8_t *)wr + offset, name, df->name_len + 1);
+        offset += df->name_len + 1;
+
+        df->value_offs = offset;
+        df->value_len = strlen(value);
+        memcpy((uint8_t *)wr + offset, value, df->value_len + 1);
+        offset += df->value_len + 1;
+
+        df++;
+    }
+
+    wr->tls = r->tls;
+    wr->nr_fields = r->fields_count;
+    wr->content_offs = offset;
+    wr->content_len = content_len = r->content_length;
+
+    read_bytes = wr->content_len;
+    if (read_bytes > NXT_WASM_MEM_SIZE - offset) {
+        read_bytes = NXT_WASM_MEM_SIZE - offset;
+    }
```

Comment:
You may want to use `read_bytes = nxt_min(wr->content_len, NXT_WASM_MEM_SIZE - offset);`